docs/12-design.md

```diff
@@ -239,33 +243,36 @@ An outline of the design of this system is shown below:
+meaning a partition split could happen before the job has run.
+
+When a compaction job finishes, it needs to update the state store to remove the references representing the input
+files in that partition, create a new reference to the output file, and update the relevant file reference counts.
+This update must be done atomically, to avoid clients that are requesting the state of the state store from seeing an 
+inconsistent view.
 
 The CDK compaction stack deploys the infrastructure that is used to create and execute compaction jobs. A compaction
-job reads in N input files and merges them into 1 or 2 output files. As the input files are all sorted by key, this
-job is a simple streaming merge that requires negligible amounts of memory. The input files are all from a single
-partition.
+job reads in N input files and merges them into 1 file. As the input files are all sorted by key, this job is a 
+simple streaming merge that requires negligible amounts of memory. The input files are all from a single partition.
 
 There are two separate stages: the creation of compaction jobs, and the execution of those jobs. Compaction jobs
 are created by a lambda that runs the class `sleeper.compaction.job.creation.CreateJobsLambda`. This lambda is
-triggered periodically by a Cloudwatch rule. The lambda iterates through each table. For each table, it queries
-the state store for information about the partitions and about the active files that do not have a job id (if a file
-has a job id it means that a compaction job has already been created for that file). It then uses a compaction
-strategy to decide what compaction jobs should be created. The compaction strategy can be configured independently
-for each table. Jobs that are created by the strategy are sent to an SQS queue.
-
-Compaction jobs are executed in containers. Currently these containers are executed in Fargate tasks but they could
+triggered periodically by a Cloudwatch rule. The lambda iterates through each table. For each table, it performs a 
+pre-splitting operation on file references in the state store. This involves looking for file references that exist 
+within non-leaf partitions, and atomically removing the original reference and creating 2 new references in the 
+child partitions. This only moves file references down one "level" on each execution of the lambda, so the 
+lambda would need to be invoked multiple times for the file references in the root partition to be moved down to the
+bottom of the tree. The lambda then queries the state store for information about the partitions and the file 
+references that do not have a job id (if a file reference has a job id it means that a compaction job has already been
+created for that file). It then uses a compaction strategy to decide what compaction jobs should be created. 
+The compaction strategy can be configured independently for each table. Jobs that are created by the strategy are sent
+to an SQS queue.
```

Comment:
It might be worth splitting this paragraph to separate the bit about reference splitting from the bit about compaction strategies.

Reply:
Thanks, I split the paragraph up, it reads a bit better now